es/es-server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2100,6 +2114,21 @@ private static void persistMetadata(
         }
     }
 
+    /**
+     * If an index/update/get/delete operation is using the special `_doc` type, then we replace
+     * it with the actual type that is being used in the mappings so that users may use typeless
+     * APIs with indices that have types.
+     */
+    private String resolveType(String type) {
```

Comment:
Do we need this? We always have only 1 type.

Maybe out of scope for this PR, but I think we could go through some of the type usages and remove them